### ISSUE TITLE
[OML for Reid] High cmc, map scores during training but average performance during inference on test set.

### ISSUE THREAD

**Issue by sujit-dn** [OML-Team/open-metric-learning#573]
Hi team, I am using OML recently for training Reid models on custom dataset. I have tried ViT, ClipVit and Resnet models from OML package. During training it seems model is getting trained well based on train loss, val loss both converge. Additionally mAP@1, 5, 10, CMC metrics both are looking good. However when inferred model on test data its doesnt seems to perform average. Any suggestions/feedback would be highly appreciated for this.

**Comment by AlekseySh** [comment]
@sujit-dn Hi, and thank you for using OML!

Sure, let's take a look at your case. Could you provide us with more details? 
How do you run you code (python/pipelines)?
What is the loss function?
You can send your code snippets, loss/metric curves and so on. It will be useful.


From what you said I may conclude that your training and validation set are from similar distribution, but your test set is kind of different. May it be the case?

**Comment by AlekseySh** [comment]
@sujit-dn , If you still work on this, please, update you package up to 3.0 and let's continue of this case

**Comment by sujit-dn** [comment]
Hi @AlekseySh, let me share answers for questions:

1. I use Pytorch lightning based training script with customized part of training and validation loop.
2. Training model with different set of combinations of loss functions - One combination consist of triplet loss only with TripletMarginMiner from PML. I've tried couple of margin values 0.1, 0.5 and 1. Got best results so far from 0.5. Second combinations consist of using Triplet + Cross Entropy loss.  
3. Dataset - I have created variations of dataset - 1) Only custom dataset 2) Custom + Market1501 (public reid dataset). 
4. Backbone - Using Resnet50, Resnet101,  Vit, ClipVit. Resnet doesnt seems to converge. Vit converge but not able to completely generalize well. ClipVit is kind of overfitting and not able to produce distinguishing features. 
5. Metrics - I got good CMC@1,5,10, map@1,5,10 scores during training for ViT and ClipViT. I evaluated ViT on holdout set and only able to get approx 70% classification acc. Holdset consists of pairs with label 1/0 for similar, dissimilar pairs.
6. Convergence - So far best results I got with Triplet loss only with 0.5. For Triplet + CE loss with ViT its kind of able to generalize now. I have created a testbench with pairs of images and labels with 1/0 for similiar/dissimilar case as ultimately I am interested in pairs classification only. Did hyperparameter search with different cut thresholds from 0.1 to 1.0 with difference of 0.1 and found best results at 0.7 as cutoff threshold. So right now its working with 0.7 as cutoff threshold.

**Comment by AlekseySh** [comment]
@sujit-dn Hi, thank you for the detailed answer.

**I think you should start with reproducing some known results to make sure that everything is all right before trying to add your data.** 

You can pick a well known paper [In Defense of the Triplet Loss for Person Re-Identification](https://arxiv.org/pdf/1703.07737) and try to get the same score with triplet loss and balance sampling on Market1501 dataset. For this you can use `BalanceSampler` and `TripletLossWithMiner` with `HardTripletsMiner` from OML. 

Note, in this paper authors showed that sometimes mode collapse may happen (when positive and negative distances are close to each other and close to zero which leads to underperforming). Probably it's your case. To fight it they suggested using soft version of triplet loss (to get it set `margin=None` in `TripletLossWithMiner`).

**Comment by AlekseySh** [comment]
One more thing: make sure you follow the Market1501 protocol for evaluation. Specifically, you should ignore photos returned from the camera id. Read more in the dataset description. Likely, there is a built-in [functionality](https://open-metric-learning.readthedocs.io/en/latest/feature_extraction/python_examples.html#handling-sequences-of-photos) in OML to apply such logic.

**Comment by sujit-dn** [comment]
Thanks for valuable feedback. One thing to I forgot to mention earlier is - I have used both OML and PML TripletLossWithMiner. For OML used BalanceSampler and TripletLossWithMiner with AllTripletsMiner with margin 0.5 . Got best results so far with this combination only. Earlier yesterday trained 4 models using combinations for ViT s16 and b16 with margin 0.5 and 1.0 on combined data (site specific + M1501). It does seems train and val loss converged and in sync. Will do further evaluation with these models and report numbers as well. Additionally I've scheduled more models training with margin 1.0 and HardTripletsMiner using site specific dataset + M1501 dataset. I'll share more results post training is done with these models. For now I've chosen to go ahead with ViT b16 and s16 architectures for experimentations. Will include ClipViT post achieving reasonable performance on dataset.

**Comment by AlekseySh** [comment]
@sujit-dn Got it. I also recommend you to try `margin = None` because sometimes it works surprisingly well. 

PS. I'm still thinking it may be a good idea to reproduce the known results first, but, of course, it's up to you :)

**Comment by sujit-dn** [comment]
![vits_16_training_graph2](https://github.com/OML-Team/open-metric-learning/assets/62235256/acc4b243-9ce4-44c6-ad8c-b507ddc896dd)
![vits_16_training_graph1](https://github.com/OML-Team/open-metric-learning/assets/62235256/513e47a0-b0c0-4685-9526-c18c01d99d3e)
Here are training graphs from recent training.

Backbone - ViT s16 with MLP head for classification
Loss - Triplet Loss + CE loss. Perspective is to train both feature extractor and classifier so it will help to get better features for images.
Margin - 0.5
Initial LR - 3e-4. 
Here is training code for reference.
[reid_trainer.py.txt](https://github.com/user-attachments/files/15806456/reid_trainer.py.txt)

Dataset - Custom dataset + M1501.

**Comment by AlekseySh** [comment]
Hi, @sujit-dn 

1. You don't normalize features in your model, so, margin=0.5 looks extremely weird. Most likely your distances are around dozens.
2. But you do normalize embeddings in the validation, which cause inconsistency.

So, I recommend you to normalize features everywhere and try margin around 0.1 or 0.2. Also try `margin=None`.

Try to watch at the graph named `active_tri` with smoothing. (Distances in only these triplets are ordered incorrectly, so, the rest of the loss values are zeros, so, you have no signal there. If number of activate triplets small it's an indicator of convergency / lack of signal).

Try HardMiner.

Try ArcFace loss.

Good luck with your experiments! :)

**Comment by sujit25** [comment]
Hi @AlekseySh, thanks for quick response.

I do normalize features in both cases as same function is getting invoked in training_step and validation_step. While training first set of models I have used 0.1 and 0.2 but it didnt performed well on hold out set. So I decided to go ahead with margin 0.5. I'll still retrain models with lower margin and let you know. Here is highlighted snap for your reference. Also right now I am using HardTripletsMiner only. Earlier trained models with AllTripletsMiner but I feel eventually it will work with HardTripletsMiner only.

<img width="873" alt="code_snippet" src="https://github.com/OML-Team/open-metric-learning/assets/8065697/d002e28d-04e0-4e4b-9834-dba0a2cf4306">

**Comment by AlekseySh** [comment]
@sujit25 oh, okay, got it

also try ArcFace and margin=None in triplet loss

**Comment by AlekseySh** [comment]
@sujit-dn Forgot to mention, don't normalize features when margin is None

this and other tips are here: https://github.com/OML-Team/open-metric-learning/tree/main/pipelines/features_extraction#tips

**Comment by sujit-dn** [comment]
Hi @AlekseySh, I did ran experiment with ViT s16 with Arcface loss from PML along with Cross Entropy loss. Parameters are default. Currently this experiment is done only on site specific dataset. 

<img width="1387" alt="vits_16_training_graph1_arcface" src="https://github.com/OML-Team/open-metric-learning/assets/62235256/7e0f399f-357a-4972-9817-3fa1b4e6a4d2">
<img width="1372" alt="vits_16_training_graph2_arcface" src="https://github.com/OML-Team/open-metric-learning/assets/62235256/3650ccd8-7549-45b8-9f75-bef4f5374589">
So it seems its converging but val acc plateaus in 70s.

**Comment by AlekseySh** [comment]
@sujit25 Sorry, I can't answer for PML's implementation of ArcFace, I only know what is in the our :) 

Still suggest you to try margin=None and wait longer than usual (turn off normalisation before)

**Comment by sujit-dn** [comment]
Hi @AlekseySh, I have interesting finding. Val accuracy is plateauing around 75% however when validated on real world data with manual QA for correct pairs its performing really well like almost 90+%. So overall it seems like some issue with my validation set. I have trained couple of models with arcface loss as well and got similar val set acc. So I'll retrain with my val set fixed. It seems like current approach of using Triplet (M=0.5)+ CE, Arcface (M=0.5) + CE is working correctly. Thanks for assistance.

**Comment by AlekseySh** [comment]
Great! That's why I said you could start with reproducing the known results. In that case you could understand that something is wrong with your own data from the very beginning. Never mind.

By the way, could we add a logo of your company (or another organization) to the "trusted by" section in the Readme?

**Comment by AlekseySh** [comment]
And can we close the issue?

**Comment by sujit-dn** [comment]
Sure. We can close this issue now.

**Comment by sujit-dn** [comment]
Sure. We can use logo from [website](https://www.adagrad.ai). If more info. required pls email me sujit@adagrad.ai.
